CARDS-1463: Computed questions should have a data type

acrowthe opened the pull request:
- Add an "entryMode" flag, with default mode "user" and secondary mode
  "computed"
- Add computed field support for most dataTypes:
  - boolean
  - date (year, date and datetime)
  - long
  - double
  - decimal
  - vocabulary
  - time (hh:mm, mm:ss)
  - text
- Add tests for each dataType to ServersideComputedTest and ComputedTest

acrowthe committed:
CARDS-1463: Computed questions should have a data type
- Add an "entryMode" flag, with default mode "user" and secondary mode
  "computed"
- Add computed field support for most dataTypes:
  - boolean
  - date (year, date and datetime)
  - long
  - double
  - decimal
  - vocabulary
  - time (hh:mm, mm:ss)
  - text
- Add tests for each dataType to ServersideComputedTest and ComputedTest

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/ComputedQuestion.jsx:
```diff
@@ -146,43 +205,108 @@ let ComputedQuestion = (props) => {
+      answerType = "String";
+      answerNodeType = "cards:ComputedAnswer";
+      break;
+  }
```
Most of this can be replaced with `answerNodeType = "cards:" + capitalizedDataType+ "Answer"`.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/ComputedQuestion.jsx:
```diff
@@ -146,43 +205,108 @@ let ComputedQuestion = (props) => {
     <Question
       defaultDisplayFormatter={isFormatted ? (label, idx) => <FormattedText>{label}</FormattedText> : undefined}
-      currentAnswers={typeof(value) !== "undefined" && value !== "" ? 1 : 0}
+      currentAnswers={typeof(displayValue) !== "undefined" && displayValue !== "" ? 1 : 0}
```
Why are we switching to `displayValue`?

sashaandjic reviewed (commented):
There is an error in the definition of the computed filed below ( highlighted `2` should be replaced with `1`) - [screenstot](https://user-images.githubusercontent.com/23243244/141379239-b7029120-7f59-4338-9f13-a42f68e81804.png)

sashaandjic commented:
BUG: Navigating off ServersideComputedTest form brings the login dialog.

acrowthe committed:
CARDS-1463: Computed questions should have a data type
- Fix type in testing form
- Fix error caused by duplicated questions in recurrent sections

acrowthe committed:
CARDS-1463: Computed questions should have a data type
Fix lint error

sashaandjic commented:
Previously reported problems are fixed, however both test forms: `Computed Field Test` and `Serverside Computed Field Test` have some `text` fields defined as `computed`. For example: 
![image](https://user-images.githubusercontent.com/23243244/141878800-42417333-800a-4619-8fef-c8bb1b6369e7.png)
instead of: 
![image](https://user-images.githubusercontent.com/23243244/141879094-5e2ba5a6-ab63-491e-b178-d874dab64b2c.png)

Also all proms forms need to be adjusted. e.g. 
![image](https://user-images.githubusercontent.com/23243244/141879047-04d2e366-c3d6-44ab-b0ad-5ebb431be8f4.png)

marta- committed:
CARDS-1463: Computed questions should have a data type
* Removed `computed` dataType from Question.json specification for the questionnaire wizard
* Updated PROMs questionnaires to the new mode of specifying computed questions
* Updated test questionnaire to the new mode of specifying computed questions

sashaandjic commented:
QUESTION: Entry mode `computed` has the following 3 display modes:
- `input`
- `formatted`
- `hidden`

The difference between `input` and `formatted` is minimal. `input` displays in gray fonts, has line below and limits the length of the field. 
Is `input` mode needed?
If not, do we than need `formatted` mode? Can't we always assume that `computed` are `formatted` and have `hidden` as a property of boolean type?

![image](https://user-images.githubusercontent.com/23243244/142127481-9cf3171b-b651-4530-bf43-aa2e6a32dc06.png)
![image](https://user-images.githubusercontent.com/23243244/142127067-716a102a-0623-463d-a2fa-71af7ef4dcd3.png)

sashaandjic commented:
Can we display the following together in the designer?
- `Data type` , `Entry mode` and `Display mode` 
- `Min answers` and `Max answers`
![image](https://user-images.githubusercontent.com/23243244/142128992-2e126b5c-a741-4079-86f1-404be02f0c96.png)

sashaandjic commented:
Not sure if this is a problem. 
When I opened the `smoking-history-v2-referral-summary` in edit mode and tried to save it without doing any changes I got this:
![image](https://user-images.githubusercontent.com/23243244/142129471-df489ca2-a145-4569-9594-845ccf6b9f06.png)
Once I put " " it stopped complaining. 
![image](https://user-images.githubusercontent.com/23243244/142129590-029241e6-20fa-4d8a-8c28-92c6b86e15e1.png)

sashaandjic commented:
Not sure about `Unsupported computed` displayed in red fonts for these 2 fields:
![image](https://user-images.githubusercontent.com/23243244/142132347-028696ef-ef3a-4b92-86ce-a1cba0c02935.png)
![image](https://user-images.githubusercontent.com/23243244/142132364-8d863a31-1725-4e64-9e23-ca5d2a183683.png)

marta- commented:
> QUESTION: Entry mode `computed` has the following 3 `Display modes`:
> 
> * `input`
> * `formatted`
> * `hidden`
> 
> The difference between `input` and `formatted` is minimal. `input` displays in gray fonts, has line below and limits the length of the field. Is `input` mode needed? If not, do we than need `formatted` mode? Can't we always assume that `computed` are `formatted` and have `hidden` as a property of boolean type?

The `input`display mode was originally the only one, since computed questions were developed before markdown integration (which allowed the introduction of the `formatted` display mode).

I don't really agree with removing it. For number answers that actually fit in an input, facilitates consistent alignment throughout the form, especially when there are units of measurement. I am, however, open to making `formatted` the default display mode instead of `input`. However, this is out of scope for 1463 and should be decided and implemented separately so as to not hold the merge of this PR.

marta- commented:
> Not sure about `Unsupported computed` displayed in red fonts for these 2 fields:

That's my bad, I forgot to update the Serverside Computed Field Test questionnaire and now the wizard doesn't understand dataType: computed. Fixing.

marta- committed:
CARDS-1463: Computed questions should have a data type
* Updated Serverside Computed Field questionnaire to the new mode of specifying computed questions

marta- commented:
> > Not sure about `Unsupported computed` displayed in red fonts for these 2 fields:
> 
> That's my bad, I forgot to update the Serverside Computed Field Test questionnaire and now the wizard doesn't understand dataType: computed. Fixing.

Done in f98ea1b .

marta- commented:
> Can we display the following together in the designer?
> 
> * `Data type` , `Entry mode` and `Display mode`
> * `Min answers` and `Max answers`

Not sure why they're scrambled like this. Please make a separate task and I'll look into it.

marta- commented:
> Not sure if this is a problem. When I opened the `smoking-history-v2-referral-summary` in edit mode and tried to save it without doing any changes I got this:
> 
> Once I put " " in the `Text` field, it stopped complaining.

That's because we're ignoring string values consisting of only white spaces: https://github.com/data-team-uhn/cards/blob/dev/modules/data-entry/src/main/frontend/src/questionnaireEditor/TextInput.jsx#L34.

Please make a task, it's an easy fix and hopefully won't have unwanted side effects (tbd by testing).

sashaandjic reviewed (approved):
tested
2 issues previously reported are now fixed:
![image](https://user-images.githubusercontent.com/23243244/142494544-afe91c35-3659-4095-9dea-c3d79e1e1011.png)

![image](https://user-images.githubusercontent.com/23243244/142494531-3060c4a0-3454-4968-b0b3-1c93bfa8d980.png)

marta- committed:
CARDS-1463: Computed questions should have a data type
* Updated ComputedQuestion component to no longer be activated by the deprecated "computed" data type

marta- committed:
CARDS-1463: Computed questions should have a data type
CARDS-1523: Better property placement in the Questionnaire designer

marta- commented:
> Can we display the following together in the designer?
> 
> * `Data type` , `Entry mode` and `Display mode`
> * `Min answers` and `Max answers`

Because of how the presence of some properties depends on the value of others, we can't always display `Data type` , `Entry mode` and `Display mode` together. However commit 8985e50 ensures that at least `Min answers` and `Max answers` are always together. It also restricts the possible values of `minAnswers` to `0` and `1` for questions with `entryMode=computed`, to match the current implementation.

sashaandjic reviewed (approved):
tested, looks much better with `min` and `max` being placed next to each other
![image](https://user-images.githubusercontent.com/23243244/142561710-6c292a6b-3fb0-4958-8f11-51ef78f0d182.png)

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/api/QuestionnaireUtils.java:
```diff
@@ -101,6 +101,15 @@
+     * Check if the given node is a Question node for a computed question.
+     *
+     * @param node the node to check, a JCR Node, may be {@code null}
+     * @return {@code true} if the node is not {@code null}, is computed and is of type {@code cards:Question},
```
```suggestion
     * @return {@code true} if the node is not {@code null}, is of type {@code cards:Question} and is computed,
```

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/ExpressionUtils.java:
```diff
@@ -86,7 +93,10 @@ static String evaluate(final Node question, final Map<String, Object> values)
-            return ValueFormatter.formatResult(result);
+            if (type == Type.DOUBLE) {
+                LOGGER.error(String.valueOf(result));
+            }
```
```suggestion
```

sdumitriu committed:
[misc] Fixed rebase issues

sdumitriu committed:
[misc] Upgrade the checkstyle configuration to avoid a bogus ClassFanOut complexity

sdumitriu committed:
[misc] Better javadoc

sashaandjic reviewed (approved):
re-tested; all good

sdumitriu committed:
CARDS-1463: Computed questions should have a data type
Optimize the code

sdumitriu committed:
[misc] Fixed typo

sashaandjic reviewed (approved):
re-tested

sdumitriu committed:
[misc] Removed leftover comments, made more variables final

sdumitriu commented on test-resources/src/main/resources/SLING-INF/content/Questionnaires/ServersideComputedTest.xml:
```diff
@@ -398,36 +666,415 @@ Expected output:
+					<value>HANCESTRO</value>
+				</values>
+				<type>String</type>
+			</property>
```
Really?

sdumitriu committed:
CARDS-1463: Computed questions should have a data type
statusFlags should be an array, not a string

sdumitriu committed:
[misc] Wrong data type for conditionalQuestionTime

sashaandjic commented:
The problem described in [CARDS-1552](https://phenotips.atlassian.net/browse/CARDS-1552) is still present.
https://user-images.githubusercontent.com/23243244/146568178-ef4bdac6-7a0f-43af-acac-ab99c83b2c6c.mp4
